modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/DistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query;
+
+import java.util.UUID;
+import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Reducer for distributed cache query.
+ */
+public interface DistributedCacheQueryReducer<T> extends CacheQueryReducer<T> {
+    /**
+     * Callback that invoked after getting a page from remote node. Checks whether it is the last page for query or not.
+     *
+     * @param nodeId Node ID of remote page.
+     * @param last Whether page is last for specified node.
+     * @return Whether page is last for a query.
+     */
+    public boolean onPage(@Nullable UUID nodeId, boolean last);
+
+    /**
+     * Loads full cache query result pages from remote nodes. It can be done for speedup operation if user invokes
+     * get() on {@link GridCacheQueryFutureAdapter} instead of using it as iterator.
+     *
+     * @throws IgniteInterruptedCheckedException If thread is interrupted.
+     */
+    public void loadAll() throws IgniteInterruptedCheckedException;
+
+    /**
+     * Callback to handle node left.
+     *
+     * @param nodeId Node ID that left a cluster.
+     * @return {@code true} if specified node runs this query.
```

Comment:
It is unclear, what result will be returned if the node has gone after the last page was received and how it should be interpreted.
Or what if the node has gone, but we already received sufficient data for finishing the query, 
e.g. query has limits, but the user app consumes data slowly. 

I think Reducer should only check internal invariants and mark node as failed.
And caller should recheck 'if the reducer is alive' right after calling this method.

Reply:
Renamed this method and made docs more clear. Now it just checks whether node is running a query in this moment. `onNodeLeft` now handles with `onError` method.